reseq_ros2/reseq_ros2/joint_publisher.py

```diff
@@ -0,0 +1,267 @@
+        current module
+        """
+        for mod in modules:
+            id = mod % 16
+
+            self.controller_pubs.append(
+                self.create_publisher(TwistStamped, f"/diff_controller{id}/cmd_vel", 10)
+            )
+
+            self.wheel_velocities.append([0.0, 0.0])
+
+            if mod in joints:
+                self.position_states.update({
+                    (mod, x): State(x + f"_{id}_joint", 0.0)
+                    for x in self.states_from_type(rc.StateType.JOINT_FEEDBACK)
+                })
+
+            if mod == end_effector:
+                self.position_states.update({
+                    (mod, x): State(x + f"_{id}_joint", 0.0)
+                    for x in self.states_from_type(
+                        rc.StateType.END_EFFECTOR_FEEDBACK
+                    )
+                })
+
+    def create_subs(self, modules: list[int], joints: list[int], end_effector: int):
+        """
+        Creates the subscription to all the feedback topics linked to Joints
+        (the ones in `rc.states`)
+        """
+        for mod in modules:
+            for topic in self.topics_from_type(rc.StateType.MOTOR_FEEDBACK):
+                self.create_subscription(
+                    Motors,
+                    f"reseq/module{mod}/{topic}",
+                    lambda msg, addr=mod, tp=topic: self.update_callback(
+                        msg, addr, tp, rc.StateType.MOTOR_FEEDBACK
+                    ), 
+                    10,
+                )
+            if mod in joints:
+                for topic in self.topics_from_type(rc.StateType.JOINT_FEEDBACK):
+                    self.create_subscription(
+                        Float32,
+                        f"reseq/module{mod}/{topic}",
+                        lambda msg, addr=mod, tp=topic: self.update_callback(
+                            msg, addr, tp, rc.StateType.JOINT_FEEDBACK
+                        ),
+                        10,
+                    )
+            if mod is end_effector:
+                for topic in self.topics_from_type(rc.StateType.END_EFFECTOR_FEEDBACK):
+                    self.create_subscription(
+                        Int32,
+                        f"reseq/module{mod}/{topic}",
+                        lambda msg, addr=mod, tp=topic: self.update_callback(
+                            msg, addr, tp, rc.StateType.END_EFFECTOR_FEEDBACK
+                        ),
+                        10,
+                    )
+
+    def broadcast_states(self):
+        """
+        Creates and publishes the `JointState` messages, based on a timer
+        of `rc.sample_time` seconds
+        """
+
+        for i, vel in enumerate(self.wheel_velocities):
+            vl = vel[0]
+            vr = vel[1]
+
+            # compute velocity of the module given the feedback velocity of its wheels
+            w = (vr-vl)/self.b
+            r = self.b/2*((vr+vl)/(vr-vl)) if (vr != vl) else 0
```

Comment:
`r` is currently unused but I think it should be used to compute `angular.z` of the `Twist` message. I guess that the angular rotation of the `Twist` is considered around the center of the module, while the one from the formulas on [Wikipedia](https://en.wikipedia.org/wiki/Differential_wheeled_robot#Kinematics_of_Differential_Drive_Robots) is centered in the IRC. However, I don't know if this makes any difference and, if so, how to account for it.

@Tizio0o0o0o Any clues?

Reply:
I think that we do not need the radius in this case and that the angular velocity is always centered in the IRC